File: Assets/LuaScripts/Api/CollectorApi.lua
local Requests = require('Common.Requests')
local Settings = require('Common.Settings')
local Util = require('Common.Util')
local rapidjson = require('rapidjson')

local apiPrefix = '/api/v1'

---@class CollectorApi
local CollectorApi = Class('CollectorApi')

---@type CollectorApi
function CollectorApi:GetInstance()
    if self.instance == nil then
        self.instance = self.New()
    end
    return self.instance
end

function CollectorApi:GetApiUrl()
    return Settings:GetInstance():GetString(Settings.KeyServerUrl)..apiPrefix
end

---Call in coroutine
function CollectorApi:Request(url, option)
    option = option or {}
    Util.StringifyTableValues(option.headers)
    Util.StringifyTableValues(option.parameters)
    Util.StringifyTableValues(option.formData)
    Util.StringifyTableValues(option.formUrlencodedDatas)
    local r = Requests.RequestSync(url, option)
    if r then
        return r
    else
        print('Request error: '..url)
    end
end

---Call in coroutine
--@param ids rapidjson.array
function CollectorApi:IllustInfo(ids)
    local s, r = pcall(function()
        local url = self:GetApiUrl()..'/illust/info'
        local body = {
            breifMode = true,
            ids = ids
        }
        local option = {
            method = 'POST',
            headers = {
                ['Content-Type'] = 'application/json'
            },
            body = rapidjson.encode(body)
        }
        local result = self:Request(url, option)
        local decode = rapidjson.decode(result.body)
        return decode
    end)
    if s then
        return r
    else
        print('IllustInfo error: '..r)
    end
end

---Call in coroutine
function CollectorApi:IllustPut(illusts)
    local s, r = pcall(function()
        local url = self:GetApiUrl()..'/illust'
        local option = {
            method = 'PUT',
            headers = {
                ['Content-Type'] = 'application/json'
            },
            body = rapidjson.encode(illusts)
        }
        local result = self:Request(url, option)
        local decode = rapidjson.decode(result.body)
        return decode
    end)
    if s then
        return r
    else
        print('IllustPut error: '..r)
    end
end

---Call in coroutine
function CollectorApi:IllustDelete(ids)
    local s, r = pcall(function()
        local url = self:GetApiUrl()..'/illust'
        local option = {
            method = 'DELETE',
            headers = {
                ['Content-Type'] = 'application/json'
            },
            body = rapidjson.encode(ids)
        }
        local result = self:Request(url, option)
        local decode = rapidjson.decode(result.body)
        return decode
    end)
    if s then
        return r
    else
        print('IllustDelete error: '..r)
    end
end

---Call in coroutine
function CollectorApi:IllustList(offsetOid)
    local s, r = pcall(function()
        local url = self:GetApiUrl()..'/illust/list'
        local body = {
            offsetOid = offsetOid
        }
        local option = {
            method = 'POST',
            headers = {
                ['Content-Type'] = 'application/json'
            },
            body = rapidjson.encode(body)
        }
        local result = self:Request(url, option)
        local decode = rapidjson.decode(result.body)
        return decode
    end)
    if s then
        return r
    else
        print('IllustList error: '..r)
    end
end

return CollectorApi

File: TsProject/src/api/collector-api.ts
import { Pamisu } from "csharp";
import { $promise, $typeof } from "puerts";
import Settings from "../common/settings";
import Util from "../common/util";
import { Illust, RequestOption } from "./entities";

const apiPrefix = '/api/v1';

export default class CollectorApi {

    private networkManager: Pamisu.Common.NetworkManager;

    private static _instance: CollectorApi;
    public static get instance(): CollectorApi {
        if (!this._instance)
            this._instance = new CollectorApi();
        return this._instance;
    }
    private constructor() {
        this.networkManager = Pamisu.Common.NetworkManager.Instance;
    }

    private get serverUrl() {
        return Settings.config.collectorServer + apiPrefix;
    }

    private async request(url: string, option: RequestOption = null) {
        option = option || {};
        const opt = Util.jsToCharp(option, $typeof(Pamisu.Common.RequestOption));
        const result = await $promise(this.networkManager.RequestAsync(url, opt));
        if (result) 
            return result;
        else 
            console.error('Request error: ' + url);
    }

    public parse(str: string): any {
        try {
            return JSON.parse(str);
        } catch (error) {
            console.error('Json parse error.');
            console.log(error);
        }
    }

    public async illustInfo(ids: number[]) {
        const url = this.serverUrl + '/illust/info';
        const body = {
            breifMode: true,
            ids
        };
        const option: RequestOption = {
            method: 'POST',
            headers: {
                ['Content-Type']: 'application/json'
            },
            body: JSON.stringify(body),
        }
        const result = await this.request(url, option);
        if (result)
            return this.parse(result.body);
    }

    public async illustPut(illusts: Illust[]) {
        const url = this.serverUrl + '/illust';
        const option: RequestOption = {
            method: 'PUT',
            headers: {
                ['Content-Type']: 'application/json'
            },
            body: JSON.stringify(illusts)
        };
        const result = await this.request(url, option);
        if (result)
            return this.parse(result.body);
    }

    public async illustDelete(ids: number[]) {
        const url = this.serverUrl + '/illust';
        const option: RequestOption = {
            method: 'DELETE',
            headers: {
                ['Content-Type']: 'application/json'
            },
            body: JSON.stringify(ids)
        };
        const result = await this.request(url, option);
        if (result)
            return this.parse(result.body);
    }

    public async illustList(offsetOid?: string) {
        const url = this.serverUrl + '/illust/list';
        const body = { offsetOid };
        const option: RequestOption = {
            method: 'POST',
            headers: {
                ['Content-Type']: 'application/json'
            },
            body: JSON.stringify(body),
        }
        const result = await this.request(url, option);
        if (result)
            return this.parse(result.body);
    }

}
